feat: add ruby plugin

steinbro opened the pull request:
- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `make lint && make test`?
- [ ] Have you added an entry to the changelog (`docs/reference/changelog.rst`)?

---
Fixes #338. The goal is to update the abandoned PR #341 as follows:

1. Don't recompile Ruby every time, i.e. allow for a ruby-deps part that can be shared by other parts
2. Add tests
3. Add docs

steinbro commented:
Implementation, tests and docs are ready for a first round of review.

cmatsuoka commented on craft_parts/plugins/ruby_plugin.py:
```diff
@@ -0,0 +1,207 @@
+    """
+
+    # NOTE: To update ruby-install version, go to https://github.com/postmodern/ruby-install/tags
+    RUBY_INSTALL_VERSION = "0.10.1"
```
We usually define this type of constant at top level, local to the module.

medubelko commented on craft_parts/plugins/ruby_plugin.py:
```diff
@@ -0,0 +1,207 @@
+
+    The desired Ruby interpreter is compiled using ruby-install.
+
+    The ruby plugin uses the common plugin keywords, plus the following ruby-
```
> common plugin keywords

A minor point, but there aren't really any common keys for plugins. Plugin keys are parallel with part keys, so they feel like they're part of the same collection, but there's no standard set, AFAIK.

medubelko commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,125 @@
+
+**Type:** list of strings
+
+List of gems to be installed.
```
```suggestion
The gems to install.
```

medubelko commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,125 @@
+
+Specifies which implementation of the Ruby interpreter the plugin
+should build. Supported values are ``ruby``, ``jruby``, ``rbx``,
+``truffleruby``, and ``mruby``.
```
```suggestion
The Ruby interpreter to compile and include. The supported interpreters are:

- ``ruby``
- ``jruby``
- ``rbx``
- ``truffleruby``
- ``mruby``
```

I don't believe we need to keep the word "implementation", but I'm not very familiar.

medubelko commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,125 @@
+**Default:** False
+
+When set to ``true``, the plugin compiles Ruby with support for the jemalloc
+memory allocator.
```
```suggestion
Whether to compile Ruby with support for jemalloc.
```

medubelko commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,125 @@
+<https://github.com/postmodern/ruby-install>` tool. If recompiling the
+interpreter for every Ruby-based part is not desired, a ``ruby-deps`` part
+can install Ruby and any gems in the build system. Other parts that use this
+interpreter must run ``after`` the ``ruby-deps`` part:
```
```suggestion
<https://github.com/postmodern/ruby-install>` tool.

If a project has multiple Ruby parts, by default each compiles and bundles its
own Ruby interpreter. If a special Ruby part named ``ruby-deps`` is defined, it
creates a shared interpreter and shared gems that the other Ruby parts can use.
For a Ruby part to use these shared files, it must list ``ruby-deps`` in its
``after`` key.
```

steinbro replied:
Thank you for a much better explanation. My attempt to adapt language from the poetry plugin reference did not turn out well.

medubelko reviewed (commented):
@steinbro I went into a little more detail than I usually do, so I hope this isn't too much to chew on. Let me know what you think.

steinbro commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,125 @@
+
+**Type:** list of strings
+
+Extra arguments to pass to the compiler when building the Ruby interpreter.
```
I would expect so, since each is an entirely different implementation with its own optional features.. There are some examples in the ruby-install README: https://github.com/postmodern/ruby-install/blob/master/README.md#synopsis

bepri commented:
Do we know which bash script is failing, exactly? My first guess is some sort of `bash`/`sh` inequality

steinbro commented:
I worked around the inscrutable syntax error by switching the integration test to compile the mruby variant of the Ruby interpreter, which uses Ruby's own Rakefile syntax instead of Makefiles. The only failing tests now are unrelated, coming from a different plugin (dotnet_v2).

lengau commented on tests/integration/plugins/test_ruby.py:
```diff
@@ -0,0 +1,107 @@
+
+
+@pytest.mark.slow
+def test_ruby_custom_flavor(new_dir, partitions):
```
Could we get a version of this test that doesn't set any configuration options?

steinbro commented:
Reworked to match the behavior of other plugins: by default, don't attempt to build an interpreter and raise an exception if none is present in the build environment. Also re-requesting docs review from @medubelko since I updated the Dependencies section of the plugin reference accordingly.

medubelko commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,128 @@
+------------
+
+The Ruby plugin needs the ``ruby`` interpreter to run Ruby programs but does not
+provision it by default, to allow flexibility in the choice of interpreter flavor
```
```suggestion
provide it by default, to give you flexibility in the choice of interpreter flavor
```

I think "provide" is a less expensive word for what we mean here.

steinbro replied:
I agree and I'll accept this suggestion, but FWIW this language came straight from the Go plugin reference.

medubelko commented on docs/common/craft-parts/reference/plugins/ruby_plugin.rst:
```diff
@@ -0,0 +1,128 @@
+Dependencies
+------------
+
+The Ruby plugin needs the ``ruby`` interpreter to run Ruby programs but does not
```
> `ruby` interpreter

Nitpick: do we really need to specify the binary name here and in the next paragraph? Doesn't "Ruby interpreter" mean the same thing?
